system/security/digisign/digisign.cpp

```diff
@@ -294,17 +235,58 @@ extern "C"
 {
 
 	//Returns static instance created from environment.conf key file settings
+    static CriticalSection staticDSMCrit;
     DIGISIGN_API IDigitalSignatureManager * staticDigitalSignatureManagerInstance()
     {
-        static CDigitalSignatureManager digitalSignatureManager;
-        return & digitalSignatureManager;
+        static Owned<IDigitalSignatureManager> dsm;
+
+#ifdef _USE_OPENSSL
+        if (nullptr == dsm.get())
```

Comment:
This is not thread safe - despite appearances - and why I pointed you at the jlib classes!

Theroetically the dsm.get() within the critical section does not need to be re-evaluated - which means that two threads could create an instance, update dsm, causing the first thread to crash since it would be using a version of the class which had just been freed by the second thread.

Reply:
If you are going to code it yourself you need to use an atomic..